### ISSUE TITLE
HLint suggests removing actually-used NamedFieldPuns extension

### ISSUE THREAD

**Issue by pbrisbin** [ndmitchell/hlint#898]
```console
% stack exec hlint -- --version
Stack has not been tested with GHC versions above 8.6, and using 8.8.2, this may fail
Stack has not been tested with Cabal versions above 2.4, but version 3.0.1.0 was found, this may fail
HLint v2.2.11, (C) Neil Mitchell 2006-2020
```

**.hlint.yaml**

```yaml
---
- ignore: {name: "Redundant do", within: spec}
- ignore: {name: "Eta reduce"}
- ignore: {name: "Use list comprehension"}
- ignore: {name: "Reduce duplication"}
- ignore: {name: "Use fmap"}  # we use classy prelude so this would be in error
- ignore: {name: "Avoid restricted function", within: Application}
- ignore: {name: "Use ."}  # commonly broken or less readable

# Custom Warnings
- warn: {lhs: mapM, rhs: traverse}
- warn: {lhs: mapM_, rhs: traverse_}
- warn: {lhs: forM, rhs: for}
- warn: {lhs: forM_, rhs: for_}
- warn: {lhs: return, rhs: pure}

# Specify additional command line arguments
- arguments:
  - -XBangPatterns
  - -XDataKinds
  - -XDeriveAnyClass
  - -XDeriveFoldable
  - -XDeriveFunctor
  - -XDeriveGeneric
  - -XDeriveLift
  - -XDeriveTraversable
  - -XDerivingStrategies
  - -XFlexibleContexts
  - -XFlexibleInstances
  - -XGADTs
  - -XGeneralizedNewtypeDeriving
  - -XLambdaCase
  - -XMultiParamTypeClasses
  - -XNoImplicitPrelude
  - -XNoMonomorphismRestriction
  - -XOverloadedStrings
  - -XRankNTypes
  - -XRecordWildCards
  - -XScopedTypeVariables
  - -XStandaloneDeriving
  - -XTypeApplications
  - -XTypeFamilies

- modules:
  - {name: [Data.Set], as: Set}
  - {name: [Data.Map], as: Map}
  - {name: [Data.HashSet], as: HashSet}
  - {name: [Data.HashMap.Strict], as: HashMap}
  - {name: [Data.Text], as: T}
  - {name: [Data.Text.Encoding], as: T}
  - {name: [Data.Text.IO], as: T}
  - {name: [Data.Text.Lazy], as: TL}
  - {name: [Data.Text.Lazy.Encoding], as: TL}
  - {name: [Data.Text.IO.Lazy], as: TL}
  - {name: [Data.ByteString], as: BS}
  - {name: [Data.ByteString.Lazy], as: BSL}
  - {name: [Data.ByteString.Char8], as: BS8}
  - {name: [Data.ByteString.Lazy.Char8], as: BSL8}
  - {name: [Data.List.NonEmpty], as: NE}
  - {name: [Data.Sequence], as: Seq}
```

**Settings.hs**

```hs
{-# LANGUAGE NamedFieldPuns #-}

module Settings where

envParseDogStatsSettings :: Env.Parser DogStatsSettings
envParseDogStatsSettings = do
  dogStatsSettingsHost <- undefined
  dogStatsSettingsPort <- undefined
  dogStatsSettingsMaxDelay <- undefined
  pure defaultSettings
    { dogStatsSettingsHost
    , dogStatsSettingsPort
    , dogStatsSettingsMaxDelay
    }
```

This is a real module from our project, but pared down to just enough to trigger this Issue. This file will of course not compile at all, but in our real project it compiles fine. 

```console
% stack exec hlint Punning.hs
Stack has not been tested with GHC versions above 8.6, and using 8.8.2, this may fail
Stack has not been tested with Cabal versions above 2.4, but version 3.0.1.0 was found, this may fail
/home/patrick/Settings.hs:1:1: Warning: Unused LANGUAGE pragma
Found:
  {-# LANGUAGE NamedFieldPuns #-}
Perhaps you should remove it.

1 hint
```

If removed, and in our real project module, I'll see:

```
/home/patrick/code/freckle/megarepo/backend/fancy-api/library/Settings.hs:201:8: error:
    Illegal use of punning for field ‘dogStatsSettingsMaxDelay’
    Use NamedFieldPuns to permit this
    |
201 |   pure defaultSettings
    |        ^
```

This seemed distinct from #767 -- apologies if it's a duplicate.

**Comment by ndmitchell** [comment]
Thanks for the report. A completely fresh issue, and one that probably got released in the last few releases when we converted that code to use the GHC AST. Great test case, and I've just pushed a fix, so will be fixed in the next release.

Unfortunately the next release will be breaking a reasonable number of things in the API (see https://twitter.com/ndm_haskell/status/1237104655550816256?s=20), and we want to break everything at once. That means we can't really release now, and we can't release for probably another week. Sorry :(

**Comment by pbrisbin** [comment]
Oh wow, thanks so much for the quick fix!
